theano/gof/opt.py

```diff
@@ -533,19 +544,64 @@ def process_node(self, fgraph, node):
             if len(node.inputs) != len(candidate.inputs):
                 continue
 
+            cand_has_assert = False
+
+            # Get input list of the candidate with assert removed
+            cand_inputs_assert_removed = []
+            for i in candidate.inputs:
+                if i.owner and isinstance(i.owner.op,
+                                          theano.tensor.opt.Assert):
+                    cand_has_assert = True
+                    cand_inputs_assert_removed.append(i.owner.inputs[0])
+                else:
+                    cand_inputs_assert_removed.append(i)
+
+            # if the current node has assert input, it should not be
+            # replaced with a candidate node which has no assert input
+            if node_has_assert and not cand_has_assert:
+                continue
+
+            # Get input list of the node with assert removed
+            if node_has_assert:
+                node_inputs_assert_removed = []
+                for i in node.inputs:
+                    if i.owner and isinstance(i.owner.op,
+                                              theano.tensor.opt.Assert):
+                        node_inputs_assert_removed.append(i.owner.inputs[0])
+                    else:
+                        node_inputs_assert_removed.append(i)
+            else:
+                node_inputs_assert_removed = node.inputs
+
             inputs_match = all(node_in is cand_in
-                               for node_in, cand_in in zip(node.inputs,
-                                                           candidate.inputs))
+                               for node_in, cand_in
+                               in zip(node_inputs_assert_removed,
+                                      cand_inputs_assert_removed))
+
             if inputs_match and node.op == candidate.op:
                 if (node, candidate) in self.blacklist:
                     # They were already tried, and there was an error
                     continue
 
-                # Schedule transfer of clients from node to candidate
-                pairs = list(zip(node.outputs, candidate.outputs))
+                # replace node with candidate
+                if not (node_has_assert or cand_has_assert):
+                    # Schedule transfer of clients from node to candidate
+                    pairs = list(zip(node.outputs,
+                                     candidate.outputs,
+                                     ['merge'] * len(node.outputs)))
+                else:
+                    new_inputs = self.get_merged_assert_input(node, candidate)
+                    new_node = node.op(*new_inputs)
+                    pairs = list(zip(node.outputs,
```

Comment:
Just add a comment that we replace node and the candidate with a new node that will have the assert.